Hw2

BaranovaElena opened the pull request:
в принципе все как в методичке - по шагам переписывала, разбиралась как оно работает, где-то по мелочи причесала

![q1Mi93sYtNs](https://user-images.githubusercontent.com/75544888/125983239-e1443400-946f-470c-96bd-ed94731f8a2d.jpg)
![d7R-t-atEmc](https://user-images.githubusercontent.com/75544888/125983243-9888a762-ae48-4c21-999f-3ac19c824bdc.jpg)

BaranovaElena committed:
добавила observer в фрагмент, в ViewModel кинула ему тост

BaranovaElena committed:
добавила View Binding в фрагмент и зачем-то в пустую активити (пусть все одинаково работает, если буду навигацию добавлять пригодится)

BaranovaElena committed:
добавила классы-состояния, View Model возвращает их вместо заглушки

BaranovaElena committed:
обработка приходящих от  ViewModel состояний - заупускаем крутилку пока состоние loading

BaranovaElena committed:
добавлена модель с заглушкой repoImpl и вывод всего на экран

BaranovaElena committed:
в модель добавлена функция выдачи двух списков погод

BaranovaElena committed:
в модель добавлена функция выдачи двух списков погод

BaranovaElena committed:
добавлен recycler с переключением двух списков по кнопке

BaranovaElena committed:
передача данных через bundle от адаптера в фрагмент с погодой, обработка клика по городу

BaranovaElena committed:
передача данных через bundle от адаптера в фрагмент с погодой, обработка клика по городу

BaranovaElena committed:
ui переделан под 2 списка, создан второй адаптер но без обработки

BaranovaElena committed:
ui переделан под 2 списка, создан второй адаптер но без обработки

BaranovaElena committed:
сделаны два списка на экране

BaranovaElena committed:
сделаны два списка на экране

BaranovaElena committed:
добавлена навигация. стартовым экраном сделала детальное отображение погоды в москве
DetailsFragment берет погоду в москве по умолчанию не из viewModel, надо исправить

BaranovaElena committed:
добавлена навигация. стартовым экраном сделала детальное отображение погоды в москве
DetailsFragment берет погоду в москве по умолчанию не из viewModel, надо исправить

BaranovaElena committed:
добавлены свои viewModel и appState для DetailsFragment
DetailsFragment берет погоду в Москве по умолчанию из viewModel

BaranovaElena committed:
добавлены свои viewModel и appState для DetailsFragment
DetailsFragment берет погоду в Москве по умолчанию из viewModel

BaranovaElena committed:
локализация

BaranovaElena committed:
темная тема

BaranovaElena committed:
рефакторинг кода:
применение let apply,
lazy для viewModel,
замена null-проверок на ?. ?: ,
сокращение лямбд,
функция-расширение для snackBar

BaranovaElena committed:
добавила сердечко на катрочку с городом, потом планирую делать список любимых городов в окне favorites

BaranovaElena committed:
добавила больше информации на экран с детальным отображением погоды

BaranovaElena committed:
hw5: сделана загрузка погоды с api (некрасиво прямо во фрагменте, посмотреть что вообще заработает)

BaranovaElena committed:
hw5: загрузка вынесена в отдельный класс, добавлены извлекаемые данные

BaranovaElena committed:
hw5: обработка ошибки загрузки

kshalnov commented on app/src/main/java/com/example/weather/model/City.kt:
```diff
@@ -0,0 +1,7 @@
+package com.example.weather.model
+
+data class City(
+    val city: String,
```
может лучше name? Чтобы потом не было city.city

BaranovaElena replied:
у меня на этапе загрузки погоды с api возникло желание вообще поменять структуру в этих дата классах
а так да, симпатичнее

kshalnov commented on app/src/main/java/com/example/weather/model/Repo.kt:
```diff
@@ -0,0 +1,6 @@
+package com.example.weather.model
+
+interface Repo {
+    fun getWeatherFromServer(): Weather
+    fun getWeatherFromLocalStorage(): Weather
```
подробно разберем на 7-8 уроке, надо один метод getWeather и несколько реализаций репозитория для разных источников данных

BaranovaElena replied:
хотела на пятой домашке этот момент с репозиториями в порядок привести, чтобы было как в заметках один интерфейс для загрузки данных и разные реализации - заглушка, база, из инета тянуть... решила сначала догнать домашки, потом хотелки делать. постараюсь исправить как можно раньше

kshalnov commented on app/src/main/java/com/example/weather/model/Weather.kt:
```diff
@@ -0,0 +1,10 @@
+package com.example.weather.model
+
+data class Weather(
+    val city: City = City(
+        "Москва",
+        55.755826,
```
в дата классе передаются конкретные значения, не локализованные строки, не надо так

BaranovaElena replied:
хотела сделать быстрый вариант загрузки города по умолчанию, сейчас уже этот костыль не нужен, удалю

kshalnov commented on app/src/main/java/com/example/weather/ui/viewmodel/MainViewModel.kt:
```diff
@@ -0,0 +1,25 @@
+    private val repoImpl: Repo = RepoImpl()
+) :
+    ViewModel() {
+    fun getLiveData() = liveDataToObserve
```
что за дата?  Что тут лежит? Из строчки вообще не понятно

BaranovaElena replied:
лежит то, что возвращает одна из реализаций AppState) попробую сделать очевиднее

kshalnov commented on app/src/main/java/com/example/weather/ui/viewmodel/MainViewModel.kt:
```diff
@@ -0,0 +1,25 @@
+import com.example.weather.model.RepoImpl
+
+class MainViewModel(
+    private val liveDataToObserve: MutableLiveData<AppState> = MutableLiveData(),
```
Почему мы передаем это в конструкторе?

BaranovaElena replied:
вообще да, нам же не нужно никакие от фрагмента аргументы подтягивать, нужно в модели создать лайфдату. спасибо, поправлю

kshalnov commented on app/src/main/java/com/example/weather/ui/viewmodel/MainViewModel.kt:
```diff
@@ -0,0 +1,25 @@
+
+class MainViewModel(
+    private val liveDataToObserve: MutableLiveData<AppState> = MutableLiveData(),
+    private val repoImpl: Repo = RepoImpl()
```
из названия переменной убрать Impl

BaranovaElena replied:
ок

BaranovaElena committed:
hw2: чистка кода: переименования, пустой конструктор viewModel, пустой конструктор City,

BaranovaElena committed:
hw3: исправления после rebase ветки hw2 с новым коммитом

BaranovaElena committed:
hw4: исправления после rebase ветки hw3 с новым коммитом

BaranovaElena committed:
hw5: привожу в порядок модель и репозитории:
DetailsFragment переделала на mvvm
viewModel не зависит от реализации репозитория WeathersRepo
загрузка из инета переместилась в WeathersRepoImplApi
weather содержит WeatherDTO, чтобы не было разногласий между загрузкой с api и заглушкой

BaranovaElena committed:
hw7-refactor-packages: перераспределила все по папкам по примеру как в чате обсуждали

BaranovaElena committed:
hw7: retrofit

BaranovaElena committed:
hw7: загрузка svg иконок погоды через GlideToVectorYou

BaranovaElena committed:
hw7: загрузка картинок городов в список через Glide

BaranovaElena committed:
hw8: добавлен экран с историей запросов, история хранится в базе Room

BaranovaElena committed:
hw9: добавлен запрос на gps, при запуске выводится погода в текущем местоположении

BaranovaElena committed:
hw10: hw3: исправила замечания к дз3
сделан интерфейс CityListsRepo, реализации которого содержат списки городов в разных локализациях. CityListFragment.kt теперь использует его вместо репозитоирия, содержащего такие списки с погодой
исправлена ошибка с использованием bundle для детального фрагмента внутри фрагмента со списком. сделан метод newInstance(city) и от фрагмента к фрагменту через mainActivity передается city

BaranovaElena committed:
hw10: сделала проверку для gps если он выключен

BaranovaElena committed:
hw10: подключена карта, переставляется маркер, по маркеру выводится название города, по кнопке "показать погоду здесь" открывается окно с погодой в отмеченной маркером точке

BaranovaElena committed:
hw10: добавлен ввод адреса, отображение координат введенного адреса

BaranovaElena committed:
hw10: навела порядок в strings.xml и dimens.xml

BaranovaElena committed:
hw11: подключила уведомления из firebase, сделала разные иконки для трех типов уведомлений

BaranovaElena committed:
hw11: сделала уведомление по нажатию иконки в action bar

BaranovaElena committed:
Merge pull request #11 from BaranovaElena/hw11-notify_btn

Hw11 - push notifications

BaranovaElena committed:
Merge pull request #10 from BaranovaElena/hw10

Hw10 - maps and geolocation

BaranovaElena committed:
Merge pull request #13 from BaranovaElena/hw9

Hw9 - content provider and location permission

BaranovaElena committed:
Merge pull request #8 from BaranovaElena/hw8

Hw8 - Room

BaranovaElena committed:
Merge pull request #7 from BaranovaElena/hw7

Hw7 - Glide

BaranovaElena committed:
Merge pull request #14 from BaranovaElena/hw7-refactor-packages

Hw7 refactor packages

BaranovaElena committed:
Merge pull request #5 from BaranovaElena/hw5

Hw5 - http

BaranovaElena committed:
Merge pull request #4 from BaranovaElena/hw4

Hw4

BaranovaElena committed:
Merge pull request #3 from BaranovaElena/hw3

Hw3
